main/src/addins/VersionControl/MonoDevelop.VersionControl/MonoDevelop.VersionControl/VersionInfoCache.cs

```diff
@@ -23,181 +23,146 @@
 
 		public VersionInfoCache (Repository repo)
 		{
 			this.repo = repo;
 		}
-
+					
 		public void ClearCachedVersionInfo (FilePath rootPath)
 		{
-			var canonicalPath = rootPath.CanonicalPath;
-
-			try {
-				fileLock.EnterWriteLock ();
-				foreach (var p in fileStatus.Where (e => e.Key.IsChildPathOf (rootPath) || e.Key == canonicalPath))
-					p.Value.RequiresRefresh = true;
-			} finally {
-				fileLock.ExitWriteLock ();
+			FileUpdateEventArgs args = null;
+			var canonicalPath = rootPath.CanonicalPath;
+
+			foreach (var p in fileStatus.Where (e => e.Key.IsChildPathOf (rootPath) || e.Key == canonicalPath)) {
+				p.Value.RequiresRefresh = true;
+
+				var a = new FileUpdateEventArgs (repo, p.Value.LocalPath, p.Value.IsDirectory);
+				if (args == null)
+					args = a;
+				else
+					args.MergeWith (a);
 			}
-
-			try {
-				directoryLock.EnterWriteLock ();
-				foreach (var p in directoryStatus.Where (e => e.Key.IsChildPathOf (rootPath) || e.Key == canonicalPath))
-					p.Value.RequiresRefresh = true;
-			} finally {
-				directoryLock.ExitWriteLock ();
+
+			foreach (var p in directoryStatus.Where (e => e.Key.IsChildPathOf (rootPath) || e.Key == canonicalPath)) {
+				p.Value.RequiresRefresh = true;
+			}
+
+			if (args != null) {
+				//	Console.WriteLine ("Notifying Status " + string.Join (", ", args.Select (p => p.FilePath.FullPath)));
+				VersionControlService.NotifyFileStatusChanged (args);
 			}
 		}
 
 		public VersionInfo GetStatus (FilePath localPath)
-		{
-			try {
-				fileLock.EnterReadLock ();
-
-				VersionInfo vi;
-				fileStatus.TryGetValue (localPath, out vi);
-				return vi;
-			} finally {
-				fileLock.ExitReadLock ();
-			}
+		{
+			fileStatus.TryGetValue (localPath, out var vi);
+
+			return vi;
 		}
 
 		public DirectoryStatus GetDirectoryStatus (FilePath localPath)
-		{
-			try {
-				directoryLock.EnterReadLock ();
-
-				DirectoryStatus vis;
-				if (directoryStatus.TryGetValue (localPath.CanonicalPath, out vis))
-					return vis;
-				return null;
-			} finally {
-				directoryLock.ExitReadLock ();
-			}
+		{
+			if (directoryStatus.TryGetValue (localPath.CanonicalPath, out var vis))
+				return vis;
+			return null;
 		}
 
 		public void SetStatus (VersionInfo versionInfo, bool notify = true)
-		{
-			try {
-				fileLock.EnterWriteLock ();
+		{
+			if (!versionInfo.IsInitialized)
+				versionInfo.Init (repo);
 
-				if (!versionInfo.IsInitialized)
-					versionInfo.Init (repo);
-				VersionInfo vi;
-				if (fileStatus.TryGetValue (versionInfo.LocalPath, out vi) && vi.Equals (versionInfo)) {
-					vi.RequiresRefresh = false;
-					return;
-				}
-				fileStatus [versionInfo.LocalPath] = versionInfo;
-			} finally {
-				fileLock.ExitWriteLock ();
-			}
+			if (fileStatus.TryGetValue (versionInfo.LocalPath, out var vi) && vi.Equals (versionInfo)) {
+				vi.RequiresRefresh = false;
+				return;
+			}
+
+			fileStatus.AddOrUpdate (versionInfo.LocalPath, versionInfo, (f, v) => versionInfo);

```

Comment:
```suggestion
			fileStatus[versionInfo.LocalPath] = versionInfo;
```

The setter internally handles concurrent scenarios.